src/graphics_handler.cpp

```diff
@@ -1,22 +1,48 @@
 #include "graphics_handler.h"
 
-GraphicsHandler::GraphicsHandler(SDL_Renderer *renderer) {
-    renderer_ = renderer;
+namespace {
+    const std::vector<GraphicPriority> DRAW_ORDER = {
+        GraphicPriority::BACKGROUND,
+        GraphicPriority::BACK,
+        GraphicPriority::MIDDLE,
+        GraphicPriority::FRONT,
+        GraphicPriority::UI,
+        GraphicPriority::OVERLAY
+    };
 }
 
-void GraphicsHandler::draw(SDL_Texture *texture, SDL_Rect texture_rect) {
-    SDL_RenderCopy(renderer_, texture, NULL, &texture_rect);
+GraphicsHandler::GraphicsHandler(SDL_Renderer *renderer)
+: renderer_(renderer)
+, draw_queue_({
+        {GraphicPriority::OVERLAY, std::vector<DrawRequest>() },
+        {GraphicPriority::UI, std::vector<DrawRequest>() },
+        {GraphicPriority::FRONT, std::vector<DrawRequest>() },
+        {GraphicPriority::MIDDLE, std::vector<DrawRequest>() },
+        {GraphicPriority::BACK, std::vector<DrawRequest>() },
+        {GraphicPriority::BACKGROUND, std::vector<DrawRequest>() }
+    })
+{ }
+
+void GraphicsHandler::draw(SDL_Texture *texture, SDL_Rect texture_rect, GraphicPriority priority) {
+    draw_queue_.at(priority).push_back(DrawRequest(texture, texture_rect, priority));
 }
 
-void GraphicsHandler::draw(SDL_Texture *texture, int x_pos, int y_pos) {
+void GraphicsHandler::draw(SDL_Texture *texture, int x_pos, int y_pos, GraphicPriority priority) {
     SDL_Rect texture_rect;
     SDL_QueryTexture(texture, NULL, NULL, &texture_rect.w, &texture_rect.h);
     texture_rect.x = x_pos;
     texture_rect.y = y_pos;
-    SDL_RenderCopy(renderer_, texture, NULL, &texture_rect);
+    this->draw(texture, texture_rect, priority);
 }
 
 void GraphicsHandler::update_screen() {
+    for(auto priority: DRAW_ORDER) {
+        for(auto const& draw_request: draw_queue_[priority]) {
+            SDL_RenderCopy(renderer_, draw_request.texture(), NULL, &draw_request.texture_rect());
+        }
+        draw_queue_.at(priority) = std::vector<DrawRequest>();
```

Comment:
Very good! I was looking around to make sure you were clearing the old stuff to avoid memory leaks.